### PR TITLE
multi: drop sqlite

### PR DESCRIPTION
**Pull request by buck54321**
### \*Database Upgrade\* - Create a backup before running this branch

Removes SQLite database. Everything is moved to `ChainDB`. ~~In fact, nearly everything that needed to be moved was put in *pgblockchain.go* for now.~~ ~~I'll probably do some more organizing before removing from draft status, though.~~ 

For full testing, you'll need https://github.com/dcrlabs/bug-free-happiness/pull/6

It has passed the following tests.

- `go test` with all combinations of `-tags` in *dcrpg*
- `apiload -a rotation` 
- simnet reorg test
- fresh DB test 

The database upgrade itself is fairly long. Took ~30 minutes on my SSD. This is because the entire blockchain must be requested again and the ticket pool evolved appropriately to populate the stats table. 

I think that @chappjc has expressed some interest in combining `ChainDB` and `ChainDBRPC`. ~~I didn't do that yet, but unless there a reason not to, I may grab that here as well.~~ done

In addition to some further organizing the *dcrpg* package, I imagine there is some optimization to be done in the main loop now that there are fewer data sources to sync.

### CODE REVIEW

File: db/dcrpg/internal/stats.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+// on height, so this table does not retain information related to orphaned
+// blocks.
+const (
+	CreateStatsTable = `
+		CREATE TABLE IF NOT EXISTS stats (
+    blocks_id INT4 REFERENCES blocks(id) ON DELETE CASCADE,
+    height INT4,
+    pool_size INT8,
+    pool_val INT8
+  );`
```
Comment (buck54321) [inline db/dcrpg/internal/stats.go]: I don't know why this happens with my editor, but I've seen this before and I don't know how to fix it. Only happens in multi-line string literals, and even then only sometimes.

File: db/dcrpg/upgrades.go [hunk 2]
```diff
@@ -242,10 +266,132 @@ func compatVersion1Upgrades(db *sql.DB, current, target DatabaseVersion, _ Block
...
+		} else {
+			dbTx.Rollback()
+		}
+	}()
+	makeErr := func(s string, args ...interface{}) error {
+		err = fmt.Errorf(s, args...)
+		return err
+	}
+	// Start with a height-ordered list of block data.
+	blockRows, err := u.db.Query(`
```
Comment (buck54321) [inline db/dcrpg/upgrades.go]: Didn't know if it's important to do a `queryContext` here or below. The function has the main context as part of the `updater`, but the queries are not particularly heavy.

File: db/dcrpg/internal/stats.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    VALUES ($1, $2, $3, $4)
+    ON CONFLICT (height)
+    DO UPDATE SET
+    blocks_id = $1,
+    height = $2,
+    pool_size = $3,
+    pool_val = $4
+  ;`
+
+	// The inner join serves to select only mainchain blocks.
```
Comment (chappjc) [inline db/dcrpg/internal/stats.go]: Linters will bitch about this comment.  I know it's silly, but since it's exported, we'll need either a blank line  between this comment or start the comment with SelectPoolInfo.

File: db/dcrpg/pgblockchain.go [hunk 4]
```diff
@@ -485,10 +445,10 @@ type DBInfo struct {
 // needed if database upgrades are required.
 func NewChainDB(dbi *DBInfo, params *chaincfg.Params, stakeDB *stakedb.StakeDatabase,
 	devPrefetch, hidePGConfig bool, addrCacheCap int, mp rpcutils.MempoolAddressChecker,
-	parser ProposalsFetcher, bg BlockGetter) (*ChainDB, error) {
+	parser ProposalsFetcher, client *rpcclient.Client, shutdown func()) (*ChainDB, error) {
```
Comment (chappjc) [inline db/dcrpg/pgblockchain.go]: Note to self: major version bump up to db/dcrpg/v5.  This plus the large DB upgrade is why I'm thinking 5.2 perhaps should become 6.0, otherwise this PR can go on hold until 5.2 commits are all done.

File: db/dcrpg/pgblockchain.go [hunk 5]
```diff
@@ -603,12 +564,17 @@ func NewChainDBWithCancel(ctx context.Context, dbi *DBInfo, params *chaincfg.Par
 		}
 
 		// Upgrades required
-		if bg == nil || reflect.ValueOf(bg).IsNil() {
+		if client == nil {
 			return nil, fmt.Errorf("a BlockGetter is required for database upgrades")
```
Comment (chappjc) [inline db/dcrpg/pgblockchain.go]: BlockGetter -> rpcclient.Client

File: db/dcrpg/pgblockchain.go [hunk 6]
```diff
@@ -633,7 +599,7 @@ func NewChainDBWithCancel(ctx context.Context, dbi *DBInfo, params *chaincfg.Par
 	case metaNotFoundErr:
 		log.Infof("Legacy DB versioning found.")
 		doLegacyUpgrade = true
-		if bg == nil || reflect.ValueOf(bg).IsNil() {
+		if client == nil {
 			return nil, fmt.Errorf("a BlockGetter is required for database upgrades")
```
Comment (chappjc) [inline db/dcrpg/pgblockchain.go]: BlockGetter -> rpcclient.Client

File: db/dcrpg/pgblockchain.go [hunk 7]
```diff
@@ -2906,9 +2912,73 @@ func (pgb *ChainDB) PurgeBestBlocks(N int64) (*dbtypes.DeletionSummary, int64, e
...
+		err = fmt.Errorf("rewind of StakeDatabase to height %d failed, "+
+			"reaching height %d instead", height, stakeDBHeight)
+		return nil, height, pgb.replaceCancelError(err)
+	}
+
 	return &summary, height, err
 }
 
+// RewindStakeDB attempts to disconnect blocks from the stake database to reach
+// the specified height. A channel must be provided for signaling if the rewind
```
Comment (chappjc) [inline db/dcrpg/pgblockchain.go]: The part about a channel for signalling can either be removed or changed to something like "A Context may be provided to allow cancellation of the rewind process."

File: db/dcrpg/pgblockchain.go [hunk 8]
```diff
@@ -3465,6 +3581,13 @@ func (pgb *ChainDB) StoreBlock(msgBlock *wire.MsgBlock, winningTickets []string,
 		// Update the best block in the meta table.
 		if err = pgb.SetDBBestBlock(); err != nil {
 			err = fmt.Errorf("SetDBBestBlock: %v", err)
+			return
+		}
+
+		// Insert the block stats. The tpi (TicketPoolInfo) should only be nil for a
+		// mainchain block if storing the genesis block.
```
Comment (chappjc) [inline db/dcrpg/pgblockchain.go]: Does that mean this will never be called for the genesis block, or is that avoided elsewhere?

File: db/dcrpg/pgblockchain.go [hunk 9]
```diff
@@ -4272,3 +4395,1479 @@ func (pgb *ChainDB) GenesisStamp() int64 {
...
+		}
+		pgb.stakeDB.SetPoolInfo(*hash, &tpis[i])
+	}
+	// for i := startHeight; i <= endHeight; i++ {
+	// 	winners, blockHash, err := db.DB.RetrieveWinners(i)
+	// 	if err != nil {
+	// 		return err
+	// 	}
+	// 	db.sDB.SetPoolInfo(blockHash)
+	// }
```
Comment (chappjc) [inline db/dcrpg/pgblockchain.go]: OK to nuke this comment.

File: db/dcrpg/pgblockchain.go [hunk 10]
```diff
@@ -4272,3 +4395,1479 @@ func (pgb *ChainDB) GenesisStamp() int64 {
...
+		block := new(exptypes.BlockInfo)
+		if data != nil {
+			block = pgb.GetExplorerBlock(data.Hash)
+		}
+		summaries = append(summaries, block)
+	}
+	return summaries
+}
+
+// Difficulty returns the difficulty.
```
Comment (chappjc) [inline db/dcrpg/pgblockchain.go]: `Difficulty` vs. `RetreiveDifficulty` is a bit confusing, and now's a good time to clarify it I think since we're already making breaking changes.

Perhaps we can make `Difficulty` into `CurrentDifficulty` and update the comment to reflect that it's retrieved from dcrd.  And `RetreiveDifficulty` would be better off without the "Retrieve" part (esp. since it's spelled incorrectly).... plus the comment about 24hrs is super confusing.

File: db/dcrpg/queries.go [hunk 11]
```diff
@@ -3428,6 +3428,69 @@ func appendMissedVotesPerWindow(charts *cache.ChartData, rows *sql.Rows) error {
...
+		}
+
+		// Converting to atoms.
+		blocks.Fees = append(blocks.Fees, uint64(fees))
+	}
+
+	return nil
+}
+
+// RetrievePoolAllValueAndSize returns all the pool value and the pool size
```
Comment (chappjc) [inline db/dcrpg/queries.go]: retrievePoolStats returns...

File: db/dcrpg/queries.go [hunk 12]
```diff
@@ -3428,6 +3428,69 @@ func appendMissedVotesPerWindow(charts *cache.ChartData, rows *sql.Rows) error {
...
+// cache.ChartUpdater.
+func retrievePoolStats(ctx context.Context, db *sql.DB, charts *cache.ChartData) (*sql.Rows, error) {
+	rows, err := db.QueryContext(ctx, internal.SelectPoolStatsAboveHeight, charts.PoolSizeTip())
+	if err != nil {
+		return nil, err
+	}
+	return rows, nil
+}
+
+// Append the result from RetrievePoolAllValueAndSize to the provided ChartData.
```
Comment (chappjc) [inline db/dcrpg/queries.go]: RetrievePoolAllValueAndSize -> retrievePoolStats

File: db/dcrpg/queries.go [hunk 13]
```diff
@@ -4096,3 +4165,306 @@ func UpdateBlockNextByNextHash(db SqlExecutor, currentNext, newNext string) erro
...
+	bd.PoolInfo.Value = dcrutil.Amount(val).ToCoin()
+	bd.PoolInfo.ValAvg = bd.PoolInfo.Value / float64(bd.PoolInfo.Size)
+	bd.Time = apitypes.TimeAPI{S: timestamp}
+	bd.PoolInfo.Winners = winners
+	bd.StakeDiff = dcrutil.Amount(sbits).ToCoin()
+	return bd, nil
+}
+
+// RetrieveDiff returns the difficulty in the last 24hrs or immediately after
+// 24hrs.
```
Comment (chappjc) [inline db/dcrpg/queries.go]: This comment is equally confusing as the other one.  I know this was already like this, but if you can think of a better explanation, I figure it's a good time to fix it.

File: db/dcrpg/sync.go [hunk 14]
```diff
@@ -380,31 +402,28 @@ func (pgb *ChainDB) SyncChainDB(ctx context.Context, client rpcutils.MasterBlock
...
-		winners := tpi.Winners
+		blockHash := block.Hash()
+
+		// // Winning tickets from StakeDatabase, which just connected the block,
+		// // as signaled via the waitChan.
+		// tpi, ok := pgb.stakeDB.PoolInfo(*blockHash)
+		// if !ok {
+		// 	return ib - 1, fmt.Errorf("stakeDB.PoolInfo could not locate block %s", blockHash.String())
+		// }
+		// winners := tpi.Winners
```
Comment (chappjc) [inline db/dcrpg/sync.go]: OK to nuke.

File: db/dcrpg/tables.go [hunk 15]
```diff
@@ -12,22 +12,31 @@ import (
...
+	{"testing", internal.CreateTestingTable},
+	{"proposals", internal.CreateProposalsTable},
+	{"proposal_votes", internal.CreateProposalVotesTable},
+	{"stats", internal.CreateStatsTable},
+}
+
+func createTableMap() map[string]string {
+	m := make(map[string]string, len(createTableStatements))
+	for _, pair := range createTableStatements {
+		m[pair[0]] = pair[1]
```
Comment (chappjc) [inline db/dcrpg/tables.go]: Just curious why these changes.

File: db/dcrpg/upgrades.go [hunk 16]
```diff
@@ -151,11 +157,20 @@ func updateSchemaVersion(db *sql.DB, schema uint32) error {
...
+	stakeDB *stakedb.StakeDatabase
+	ctx     context.Context
+}
+
 // UpgradeDatabase attempts to upgrade the given sql.DB with help from the
 // BlockGetter. The DB version will be compared against the target version to
 // decide what upgrade type to initiate.
-func UpgradeDatabase(db *sql.DB, bg BlockGetter) (bool, error) {
-	initVer, upgradeType, err := versionCheck(db)
+func UpgradeDatabase(u *updater) (bool, error) {
```
Comment (chappjc) [inline db/dcrpg/upgrades.go]: For this and the other functions that now take an `*updater`, I might have made the updater a receiver instead.  i.e. `func (u *updater) Upgrade() (bool, error) {`

File: db/dcrpg/upgrades.go (hunk 2 above)
Comment (chappjc) [reply db/dcrpg/upgrades.go]: Nah, I'd actually say we don't want it cancellable since it could destroy the DB if half finished.  Same reason why I din't prefer to use the context in the important write queries.

File: db/dcrpg/upgrades.go [hunk 17]
```diff
@@ -242,10 +266,132 @@ func compatVersion1Upgrades(db *sql.DB, current, target DatabaseVersion, _ Block
...
+		poolInfo := sDB.PoolInfoBest()
+		if poolInfo == nil {
+			return makeErr("PoolInfoBest error encountered")
+		}
+		// Insert rows.
+		_, err = statsStmt.Exec(id, height, poolInfo.Size, int64(poolInfo.Value*dcrToAtoms))
+		if err != nil {
+			return makeErr("insert Exec: %v", err)
+		}
+		_, err = winnersStmt.Exec(pq.Array(poolInfo.Winners), hashStr)
```
Comment (chappjc) [inline db/dcrpg/upgrades.go]: Looks like you got the pq.Array solution here, but the comment above on line 345.

BTW, I believe you can be more explicit with `pq.StringArray`, but [pq sorts it out with a type switch](https://github.com/lib/pq/blob/78223426e7c66d631117c0a9da1b7f3fde4d23a5/array.go#L41).

File: go.sum [hunk 18]
```diff
@@ -165,6 +165,7 @@ github.com/decred/dcrd/wire v1.1.0 h1:G+3CugtxNbToUN8RKWqm74yLfzJJ2BKMOr2RgWc4Ty
 github.com/decred/dcrd/wire v1.1.0/go.mod h1:/JKOsLInOJu6InN+/zH5AyCq3YDIOW/EqcffvU8fJHM=
 github.com/decred/dcrd/wire v1.2.0 h1:HqJVB7vcklIguzFWgRXw/WYCQ9cD3bUC5TKj53i1Hng=
 github.com/decred/dcrd/wire v1.2.0/go.mod h1:/JKOsLInOJu6InN+/zH5AyCq3YDIOW/EqcffvU8fJHM=
+github.com/decred/dcrdata v2.1.3+incompatible h1:DLzrqgOOgHwrUSbuOVDAUzfQePhCGl9UsjEgZI1D0Kg=
```
Comment (chappjc) [inline go.sum]: pls remove.  Usually creeps in during dev.

File: db/dcrpg/upgrades.go (hunk 17 above)
Comment (chappjc) [reply db/dcrpg/upgrades.go]: `InsertBlock` was doing the sprintf with some other text arrays (and uint arrays), so perhaps I'm mistaken about something.

File: db/dcrpg/internal/blockstmts.go [hunk 19]
```diff
@@ -53,12 +54,12 @@ const (
...
 		time, nonce, vote_bits, voters,
 		fresh_stake, revocations, pool_size, bits, sbits,
-		difficulty, stake_version, previous_hash, chainwork)
+		difficulty, stake_version, previous_hash, chainwork, winners)
 	VALUES ($1, $2, $3, $4, $5, $6,
 		$7, $8, %s, %s, $9, %s, %s,
 		$10, $11, $12, $13,
 		$14, $15, $16, $17, $18,
-		$19, $20, $21, $22) `
+		$19, $20, $21, $22, %s) `
```
[the remark this answers is not in the document]
Comment (chappjc) [reply db/dcrpg/internal/blockstmts.go]: I notice we're doing sprintf formatting for some test arrays and int arrays, so maybe there's a reason I don't recall.

File: db/dcrpg/tables.go (hunk 15 above)
Comment (buck54321) [reply db/dcrpg/tables.go]: Since `stats` has a foreign key constraint, I needed to create the tables in order.

File: go.sum (hunk 18 above)
Comment (buck54321) [reply go.sum]: every. damn. time. on my Window laptop. I've gotten into a habit of checking, but obviously not well enough.

File: db/dcrpg/internal/blockstmts.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (buck54321) [reply db/dcrpg/internal/blockstmts.go]: Yep. I'm using `pq.Array` in other places, just didn't want to break the pattern here. Happy to try switching everything to `pq.Array` or `pq.StringArray` if desired.

[the remark this answers is not in the document]
Comment (chappjc) [reply db/dcrpg/internal/blockstmts.go]: Yeah, unless it looks like a big undertaking, I think it would be worth it.  I have a feeling we won't be revisiting this code too often.

File: db/dcrpg/tables.go (hunk 15 above)
Comment (chappjc) [reply db/dcrpg/tables.go]: Ahhh, makes sense.

File: db/dcrpg/pgblockchain.go (hunk 8 above)
Comment (buck54321) [reply db/dcrpg/pgblockchain.go]: Hmm. I don't know where I got that from. The `TicketPoolInfo` is not `nil` for the genesis block.

File: db/dcrpg/internal/common.go [hunk 20]
```diff
@@ -32,41 +27,3 @@ const (
...
-		buffer.WriteString(`'` + txt + `', `)
-	}
-	buffer.WriteString("]")
-
-	return buffer.String()
-}
-
-func makeARRAYOfBIGINTs(ints []uint64) string {
-	if len(ints) == 0 {
-		return "'{}'::BIGINT[]" // cockroachdb: "ARRAY[]:::BIGINT[]"
```
Comment (chappjc) [inline db/dcrpg/internal/common.go]: Oh, I wonder if I was doing this string gen stuff because cockroachdb sucks with the lib/pq driver.  Ah well, maybe cockroach will be more compatible someday.  The right way to do this is how you got it now, with pq.Array/etc.

**Comment by chappjc** [comment]
Would you also please remove `SQLT=debug` from sample-dcrdata.conf and dcrdata-simnet.conf?

**Comment by chappjc** [comment]
All good with a fresh resync:

```
[INF] PSQL: Avg. speed: 1434 tx/s, 5263 vout/s
```

Although I miss a report at this stage (before indexing etc.) that says height and elapsed time.

**Comment by chappjc** [comment]
Upgrade went smoothly in 12 minutes:

```
2019-08-16 13:58:42.522 [INF] PSQL: DB schema version 1.1.0 upgrading to version 1.2.0
2019-08-16 13:58:42.522 [INF] PSQL: Performing database upgrade 1.1.0 -> 1.2.0
2019-08-16 13:58:42.550 [INF] SKDB: Loading ticket pool DB. This may take a minute...
2019-08-16 13:58:42.552 [INF] SKDB: badger: All 0 tables opened in 0s
2019-08-16 13:58:42.553 [DBG] SKDB: badger value log not rewritten (OK).
2019-08-16 13:58:42.553 [INF] SKDB: Loading all ticket pool diffs...
2019-08-16 13:58:42.553 [INF] SKDB: Successfully loaded 0 ticket pool diffs
2019-08-16 13:58:42.557 [INF] SKDB: Creating new stake DB.
2019-08-16 13:58:42.558 [DBG] SKDB: Initialized new stake db.
2019-08-16 13:58:42.558 [INF] SKDB: Advancing ticket pool DB to tip via diffs...
2019-08-16 13:58:42.558 [INF] SKDB: Pre-populating live ticket cache and computing pool value...
2019-08-16 13:58:42.558 [INF] PSQL: Processing blocks 0 - 9999
2019-08-16 13:59:02.566 [INF] PSQL: Processing blocks 10000 - 19999
2019-08-16 13:59:20.509 [INF] PSQL: Processing blocks 20000 - 29999
2019-08-16 13:59:38.938 [INF] PSQL: Processing blocks 30000 - 39999
2019-08-16 13:59:57.470 [INF] PSQL: Processing blocks 40000 - 49999
2019-08-16 14:00:16.433 [INF] PSQL: Processing blocks 50000 - 59999
2019-08-16 14:00:35.677 [INF] PSQL: Processing blocks 60000 - 69999
2019-08-16 14:00:55.103 [INF] PSQL: Processing blocks 70000 - 79999
2019-08-16 14:01:14.514 [INF] PSQL: Processing blocks 80000 - 89999
2019-08-16 14:01:35.288 [INF] PSQL: Processing blocks 90000 - 99999
2019-08-16 14:01:56.180 [INF] PSQL: Processing blocks 100000 - 109999
2019-08-16 14:02:17.116 [INF] PSQL: Processing blocks 110000 - 119999
2019-08-16 14:02:38.238 [INF] PSQL: Processing blocks 120000 - 129999
2019-08-16 14:03:00.541 [INF] PSQL: Processing blocks 130000 - 139999
2019-08-16 14:03:23.167 [INF] PSQL: Processing blocks 140000 - 149999
2019-08-16 14:03:45.578 [INF] PSQL: Processing blocks 150000 - 159999
2019-08-16 14:04:10.042 [INF] PSQL: Processing blocks 160000 - 169999
2019-08-16 14:04:32.820 [INF] PSQL: Processing blocks 170000 - 179999
2019-08-16 14:04:55.176 [INF] PSQL: Processing blocks 180000 - 189999
2019-08-16 14:05:17.803 [INF] PSQL: Processing blocks 190000 - 199999
2019-08-16 14:05:41.413 [INF] PSQL: Processing blocks 200000 - 209999
2019-08-16 14:06:05.344 [INF] PSQL: Processing blocks 210000 - 219999
2019-08-16 14:06:27.689 [INF] PSQL: Processing blocks 220000 - 229999
2019-08-16 14:06:48.997 [INF] PSQL: Processing blocks 230000 - 239999
2019-08-16 14:07:11.072 [INF] PSQL: Processing blocks 240000 - 249999
2019-08-16 14:07:33.432 [INF] PSQL: Processing blocks 250000 - 259999
2019-08-16 14:07:56.180 [INF] PSQL: Processing blocks 260000 - 269999
2019-08-16 14:08:17.460 [INF] PSQL: Processing blocks 270000 - 279999
2019-08-16 14:08:39.116 [INF] PSQL: Processing blocks 280000 - 289999
2019-08-16 14:09:00.639 [INF] PSQL: Processing blocks 290000 - 299999
2019-08-16 14:09:22.046 [INF] PSQL: Processing blocks 300000 - 309999
2019-08-16 14:09:43.151 [INF] PSQL: Processing blocks 310000 - 319999
2019-08-16 14:10:04.623 [INF] PSQL: Processing blocks 320000 - 329999
2019-08-16 14:10:25.520 [INF] PSQL: Processing blocks 330000 - 339999
2019-08-16 14:10:46.418 [INF] PSQL: Processing blocks 340000 - 349999
2019-08-16 14:11:07.269 [INF] PSQL: Processing blocks 350000 - 359999
2019-08-16 14:11:24.718 [DBG] PSQL: meta height 356568 / blocks height 356568
```

File: db/dcrpg/sync.go [hunk 21]
```diff
@@ -159,6 +160,30 @@ func (pgb *ChainDB) SyncChainDB(ctx context.Context, client rpcutils.MasterBlock
...
 	}
 
+	// Check and report heights of the DBs. dbHeight is the lowest of the
+	// heights, and may be -1 with an empty DB.
+	stakeDBHeight := int64(pgb.stakeDB.Height())
+	if lastBlock < -1 {
+		panic("invalid starting height")
+	}
+
+	log.Info("Current best block (chain server):    ", lastBlock)
```
Comment (chappjc) [inline db/dcrpg/sync.go]: This one is not `(chain server)`, but the postgresql db height, whatever we want to call that.
But also logging `nodeHeight` around here would be good too.
